### DISCUSSION TITLE
Can you block restrict boundary conditions?

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#23658] [Q&A General] [answered]
Hello all,

I am wondering if it is possible to block restrict a boundary condition - or indeed, if it makes any sense to?

The reason for doing this is related to [#23484](https://github.com/idaholab/moose/discussions/23484):

I am trying to implement the same conditions described by the `CoupledPenaltyInterfaceDiffusion` interface kernels, instead using custom BCs which inherit from `IntegratedBC`, and are manually applied to what would be `variable` and `neighbour_variable` if using an interface kernel.

I think my issue could be because I cannot differentiate between each side of the interface, and the `neighbour_variable` is defined on both sides. 

I am concerned that without some kind of block restriction, I am not actually setting the element residuals on each side of the interface to be equal through my custom BCs, and the BC handling `neighbour_variable` is actually contributing to the element residuals on both sides of the boundary. If so, this would prevent the residuals from being equal, no longer implying the continuity of flux as described in the linked thread.

Is this a legitimate concern and if so, is there a way to block restrict boundary conditions so the contribution of `neighbour_variable` is only applied on one side?

Thanks!

**Comment by GiudGiud** [comment, marked as answer]
If the neighbour variable is only present on one side, the BC will only contribute to its "variable=" parameter so it will only contribute on one side

Why is the neighbor_variable defined on both sides?

A BC inside a variable's domain is not normal. I think you should look at interface kernels at this point

**Reply by Eilloo** [reply]
Thanks for the suggestions,

neighbour_variable in this case (fsi) is displacement, which exists across the whole mesh to enable the mesh to deform as in the fsi example input file. 
I did think that splitting this into disp_solid and disp_fluid, enforcing an equalValueBoundaryConstraint at the interface, and combining back into a single 'disp' aux variable across the whole mesh could work. However, it seems that 'disp' must be a coupled variable, not an aux variable, for the mesh displacements to be applied.

The problem with interface kernels is that I am looking to pair a vector variable on one side of the interface, with three field variables representing components of a vector on the other side (or two for 2D, as in my current test case). From looking at the interface kernel templates, I think that accommodating two different variable types in this way becomes very messy and is worth avoiding if possible

The aforementioned vector is fluid velocity, which is required to be a vector variable for the stabilisation kernels to work. Alternatively, I could try rewriting all the bits which contribute to these terms for individual field variables. This sounds more plausible, but still very involved for what I am trying to achieve.

Since it was mentioned in the above thread that the interface kernel can be thought of as an integrated BC, I thought a simpler approach was to try and manually apply the appropriate conditions using custom BCs on each side of the interface.

Hopefully this explains my current thinking - if there might be a better way, I am very open to other options!

**Reply by GiudGiud** [reply]
What is this BC on displacement?
It seems odd that it happens inside the domain.
Are you forcing the discplacement of the solid?

**Reply by Eilloo** [reply]
The displacement of the solid should be forced by the movement/pressure of the fluid.

This means that we want to ensure the continuity of stress between the fluid and solid. 
From my understanding of lindsayad's explanation [here](https://github.com/idaholab/moose/discussions/23484), this means setting the natural BC (ie, the residual) for the solid displacement, and for the fluid velocity (where pressure is integrated by parts) to be the same value (although one will be negative due to the normals being consistent in direction at the boundary).

Just like the interface kernel, I am setting that value to be some penalty(someVar - someOtherVar), where I want someVar and someOtherVar to be equal at the interface. In this case, these are chosen to be the fluid and solid velocities.

So the boundary condition itself is the continuity of stress, achieved by applying the same residual for disp in the solid domain, and velocity in the fluid domain.

**Reply by lindsayad** [reply]
I honestly think the cleanest way is to try and figure out how to write an interface kernel that can be applied with the vector variable on one side and the scalar field variables on the other

**Reply by lindsayad** [reply]
You could create an issue asking for an analog of `PenaltyInterfaceDiffusion` that works with interfacing vector/scalar-field variables. I would be willing to tackle that issue if you open it

**Reply by Eilloo** [reply]
Okay, thanks @lindsayad, will do!

In the meantime I think I've been getting closer to making the above work, even if it is a bit of a short-term, hacky workaround.
Stopping me at the moment is an error stating that my BC depends on a variable that "...does not appear to be defined on (all of) boundary 'interface'". In this case, A variable on the solid side on which the residual of the fluid velocity depends.

It seems that the sideset created by 'SideSetsBetweenSubdomainsGenerator' only belongs to one subdomain, rather than being shared? I can only access variables belonging to the block I have specified as 'primary_block'.

Is this expected - and if so, is there any way to have a sideset be shared by both blocks?

A sidenote which could be of interest to you: I may have stumbled upon a bug:
The error only appears when I run the file on 1 or 2 nodes. Any more than that, I see no error message and the simulation attempts to run normally.
The mesh is very small, so I wonder if the sideset is somehow being spread across too many processors to determine that there is a problem... in any case, the behaviour seems unusual so I thought I should mention it!


Thanks again for all the insight!

**Reply by lindsayad** [reply]
> A sidenote which could be of interest to you: I may have stumbled upon a bug:
> The error only appears when I run the file on 1 or 2 nodes. Any more than that, I see no error message and the simulation attempts to run normally.
> The mesh is very small, so I wonder if the sideset is somehow being spread across too many processors to determine that there is a problem... in any case, the behaviour seems unusual so I thought I should mention it!

If you have a Minimal Working Example (MWE) input file that only uses MOOSE, then it would be great if you could open an issue with with that input. We do indeed want the same behavior regardless of the number of processes being used 

> It seems that the sideset created by 'SideSetsBetweenSubdomainsGenerator' only belongs to one subdomain, rather than being shared? I can only access variables belonging to the block I have specified as 'primary_block'.
>
> Is this expected - and if so, is there any way to have a sideset be shared by both blocks?

Sidesets are fundamentally "sided". They do logically belong to one block of elements or another. If you want the sideset associated with the other subdomain, then I would just create another `SideSetsBetweenSubdomainsGenerator` that has the `primary_block` and `paired_block` parameters switched. Then you'll have two sidesets that are collinear but are logically associated with different subdomain sides

**Reply by Eilloo** [reply]
Sounds good - I'll figure out what the MWE is for this and raise an issue.

Regarding sidesets, I see... I did try and think of a workaround using two sidesets like this, but the issue is I need to be able to access the variables on both sides of the boundary in order to calculate the residual. Perhaps I'm finding out why interfaces exist in the first place!

My latest workaround attempt is to us an`InterfaceUserObject`, which should have access to variables on both sides of the boundary.
The idea is to have a method in the user object which pulls out the required variable values for a given quadrature point. This gets called from the BC, passing `_qp` and away we go. 
The problem is, if I try and access something like: `neighborVar[qp]` in the user object, we segfault. 'neighbourVar' is a `coupledNeighborValue`.

I realise that _qp's are local (here, printing them out always gives 0 and 1; presumably the ends of each side, given everything is first order) so I'm thinking I need to give the user object some global qp, or at make sure it's looking at the correct element/side? It doesn't complain about the 'normal' (non-neighbor) variable though...

Does this sound possible, or is the whole approach a no-go?

**Reply by lindsayad** [reply]
Your sideset that you pass to the `boundary` parameter of your `InterfaceUserObject` should be associated with the `var` side and consequently the `neighborVar` should be on the "other" side. If your `neighborVar` is associated with the same side as your sideset, then I believe you could get segmentation faults.

If you believe you're not violating the above conditions, then it sounds like we have some bugs associated with the `InterfaceUserObject` which would be a little surprising but obviously not impossible. 

Standard `_qp` indexing should work given the conditions I outlined in the first paragraph.

**Reply by Eilloo** [reply]
Okay, so I think there is a small update on this worth mentioning - apologies for the delay!

I moved away from trying to use BCs in conjunction with InterfaceUserObjects, instead going back to the interface kernels system (more on this below). I suspect I was doing something wrong before, but if I get time I'll try and look over it again to make sure there aren't any bugs - I also suspect a bug is the less likely cause!

My new question is about vector test functions - I'll do my best to explain the updated context below:

I realised that there might be a simpler workaround using interface kernels:
The idea is to set up two interface kernels - one vector, one field, each of which contributes to the residual _only_ on the 'Element' side. The neighbour side is set to zero.
In the input file, the interface boundary is defined twice, such that we have a version with each side as the 'primary' block. This lets us use the appropriate interface kernel from each subdomain, given we only ever contribute to the 'Element' side.
On the other (neighbour) side for each subdomain is a dummy variable of the correct type, set to zero with only a nullKernel acting on it. These enable us to call each interface kernel without MOOSE complaining about the neighbour variable being a different type.
Finally, we have the residual contributions:
- For the field variable kernel, we accept the vector variable on the neighbour side as a coupled neighbour variable, along with a 'component' parameter referring to x, y, or z. We extract the appropriate component from the vector variable, and calculate the residual in exactly the same way as for one side of the existing `CoupledPenaltyInterfaceDiffusion' interface kernel.
- For the vector variable, we accept all three (x, y, and z) neighbour coupled field variables, and calculate the individual penalties associated with each. These components are assembled in a vector, which is subsequently multiplied by the vector variable test function to give us our residual for that side of the interface.

I would expect this setup to do exactly the same thing as the 'CoupledPenaltyInterfaceDiffusion' FSI example code; however, I can't seem to make it work. It did occur to me that the normals might be defined in the opposite direction on one side, but flipping the sign didn't help.

I now wonder if the test functions for vector variables are different to the case where you had three field variables? The assembly of a vector of penalties, multiplied by a vector test function seems to be the most obvious deviation from the existing interface kernels.

It's also worth noting that I have made these interface kernels AD, whereas the originals were hand-coded. I am under the impression that moving from non-AD to AD should be 'safe', since the jacobians will be exact? Perhaps I am mistaken here though, and this is what's causing problems.

Hopefully the above makes some sense. If there's any other logic flaws which jump out, I'm all ears - this would be really useful to crack, as then FSI becomes possible at high Reynolds numbers and so on :)

**Reply by lindsayad** [reply]
This is code that shows how the `LAGRANGE_VEC` shape functions are computed in 2D
```c++
// 2-D
template <> RealGradient FE<2,LAGRANGE_VEC>::shape(const ElemType type, const Order order,
                                                   const unsigned int i, const Point & p)
{
  Real value = FE<2,LAGRANGE>::shape( type, order, i/2, p );

  switch( i%2 )
    {
    case 0:
      return libMesh::RealGradient( value );

    case 1:
      return libMesh::RealGradient( Real(0), value );

    default:
      libmesh_error_msg("i%2 must be either 0 or 1!");
    }

  //dummy
  return libMesh::RealGradient();
}
```
You can see that for `_i` even, the shape functions will be like `RealVectorValue(scalar_lagrange_test[_i / 2], 0, 0)` and for `_i` odd, the shape functions will be like `RealVectorValue(0, scalar_lagrange_test[_i / 2], 0)`. Does that help?

**Reply by Eilloo** [reply]
Perfect, that's exactly what I was looking for. I've managed to find the relevant petsc docs from this, and my understanding is this:

For a vector variable, 'i' is indexed something like:
i=0: x_variable, shape function 1
i=1: y_variable, shape function 1
i=2: x_variable, shape function 2
i=3: y_variable, shape function 2
...
And so on. For instance, in 2D, with quad elements, we would have i=0-->7, corresponding to the 4 shape functions for each of the two variables. Hence, integer division by 2 (or 3 for 3D) gives us the appropriate test functions.

At first I was confused by why we are then taking the gradient; however, I now think that 'RealGradient' is simply a data type where we can store the appropriate vector of test functions, rather than calculating any gradient.

Returning to the context of CoupledPenaltyInterfaceDiffusion', I believe that a line like this in a vector interface kernel:

`residual = test[_i][_qp] * RealVectorValue(penalty_x, penalty_y)`

Is indeed the same as a line like the following which is called from multiple instances (var_x and var_y) of a scalar interface kernel:

`residual = test[_i][_qp] * penalty_component`

With a little more confidence in the kernels, testing a few other cases and petsc options revealed that PJFNK converges to something reasonable (albeit slowly), but NEWTON does not. 

I have made everything AD, so I think the jacobian should not be a problem?

The condition number is quite large (e+5-e+9 ish) using timeDerivative kernels for dummy variables instead of nullKernels (nullKernels caused the condition number to shoot up to e+30!)

`svd-monitor` also reports that there are no singular values that are (nearly) zero.

The linear residuals converge immediately (as expected using `lu` and `pc_factor_shift_type = nonzero`), but the nonlinear residual appears to show no change after one step, then the solver reports that it has diverged.

I am testing everything using the example flat channel input file, which gives me some confidence in the boundary conditions... 
Given the above, where would you suggest I look for reasons that NEWTON does not converge, when PJFNK does?

**Reply by lindsayad** [reply]
> At first I was confused by why we are then taking the gradient; however, I now think that 'RealGradient' is simply a data type where we can store the appropriate vector of test functions, rather than calculating any gradient.

exactly correct

> I am testing everything using the example flat channel input file, which gives me some confidence in the boundary conditions...
> Given the above, where would you suggest I look for reasons that NEWTON does not converge, when PJFNK does?

Do you have a git branch I could pull and do some testing?

**Reply by Eilloo** [reply]
I've added the new kernels and input files to a new app in a git repo here:

https://github.com/Eilloo/FlatChannelFSI_VectorNS.git

Let me know if that doesn't work for any reason!

**Reply by lindsayad** [reply]
I tried compiling this with the `devel` branch of MOOSE and couldn't compile

**Reply by Eilloo** [reply]
Ah, sorry - my bad, it should work now!

**Reply by lindsayad** [reply]
I just pushed a commit. You had `value()` in your AD interface kernels. That blows away all the AD computations for the Jacobian. You should never use that function

**Comment by Eilloo** [comment]
Aaah, it's always something small - that's good to know, I'd misunderstood what `value()` does from the docs.
That has done the trick - thanks a lot for taking a look! :)

**Reply by lindsayad** [reply]
You're welcome!
